### PR TITLE
Add an oplog to DependencyGraph to avoid copying it

### PR DESCRIPTION
**Pull request by segiddins**
Along with other performance changes made in bundler, this is good for up to a 3x drop in object allocations and a 3x speed up. Probably can optimize further, but I'm just too excited to keep this to myself much longer 🎉 

- [x] CHANGELOG
- [x] Docs
- [x] Performance tuning
- [x] Get rid of metaprogramming

@alloy @indirect

### CODE REVIEW

**Comment by indirect** [comment]
:tada:

File: lib/molinillo/dependency_graph/log.rb [hunk 1]
```diff
@@ -0,0 +1,135 @@
+require 'molinillo/dependency_graph/action'
+
+module Molinillo
+  class DependencyGraph
+    class Log
+      def self.action(name__, parameters = nil, &blk)
+        name__ = name__.to_sym
+        cls = Class.new(Action) do
+          define_singleton_method(:name) { name__ }
```
Comment (segiddins) [inline lib/molinillo/dependency_graph/log.rb]: `define_singleton_method` is no good because of 1.8.7

File: lib/molinillo/dependency_graph/log.rb [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+            parameters ||= method.parameters.map(&:last)
+            module_eval <<-EOS
+              #{parameters.map { |param| "attr_reader(:#{param});" }.join}
+              define_method(:initialize) do |#{parameters.join(', ')}|
+                #{parameters.map { |p| "@#{p} = #{p};" }.join}
+                @args = [#{parameters.join(', ')}]
+              end
+            EOS
+          end
+          module_exec(&blk)
```
Comment (alloy) [inline lib/molinillo/dependency_graph/log.rb]: If for whatever reason you really need to eval a string, then it's best for debuggability that you include the file and line where this was defined.

**Comment by alloy** [comment]
Very awesome! :clap:

**Comment by orta** [comment]
Ace

**Comment by segiddins** [comment]
Might just get rid of the metaprogramming -- it turns out, there are way fewer primitive operations than I thought there would be!

**Comment by segiddins** [comment]
Also should probably profile allocations and see if there's any low-hanging fruit there while we're at it

**Comment by segiddins** [comment]
@indirect can has review?

**Comment by indirect** [comment]
doo ittttt r+

**Pull request merged**